pkg/scheduler/objects/queue.go

```diff
@@ -1017,17 +1016,18 @@ func (sq *Queue) IncAllocatedResource(alloc *resources.Resource, nodeReported bo
 	sq.Lock()
 	defer sq.Unlock()
 	// all OK update this queue
-	sq.allocatedResource = newAllocated
+	sq.allocatedResource = resources.Add(sq.allocatedResource, alloc)
 	sq.updateAllocatedResourceMetrics()
 	return nil
 }
 
+// allocatedResFits adds the passed in resource to the allocatedResource of the queue and checks if it still fits in the
+// queues' maximum. If the resource fits it returns true otherwise false.
 // small helper method to access sq.maxResource+sq.allocatedResource and avoid Clone() call
-func (sq *Queue) allocatedResFits(res *resources.Resource) (bool, *resources.Resource) {
+func (sq *Queue) allocatedResFits(res *resources.Resource) bool {
 	sq.RLock()
 	defer sq.RUnlock()
-	newAllocated := resources.Add(sq.allocatedResource, res)
```

Comment:
I'm thinking about this...

This is the error message which Craig reproduced:

```
2024-07-26T14:42:30.364Z    DPANIC    core.scheduler.application    objects/application.go:1496    queue update failed unexpectedly    {"error": "allocation (map[pods:1]) puts queue 'root' over maximum allocation (map[[devices.kubevirt.io/kvm:0](http://devices.kubevirt.io/kvm:0) [devices.kubevirt.io/tun:0](http://devices.kubevirt.io/tun:0) [devices.kubevirt.io/vhost-net:0](http://devices.kubevirt.io/vhost-net:0) ephemeral-storage:744815809367 hugepages-1Gi:0 hugepages-2Mi:0 memory:1115240587264 [nvidia.com/gpu:0](http://nvidia.com/gpu:0) pods:660 vcore:320000]), current usage (map[memory:314572800 [nvidia.com/gpu:1](http://nvidia.com/gpu:1) pods:23 vcore:200])"}
```

It says `[nvidia.com/gpu:0]` and this is printed from `sq.allocatedResource`, which means that this resource type is actually present. That is, if we want to prevent this bug, this change won't solve it because both `AddOnlyExisting()` and `FitInMaxUnder()` will consider the gpu type. Am I missing something here?

I haven't run any code just trying to have a good understanding simply by looking at it :)

Reply:
Yes, you can absolutely end up with greater usage than capacity. This is going to become even more of an issue in the future with mutable pod resources. In this case, the issue was discovered when using GPU nodes. Restarting the NVidia GPU operator causes the node to stop reporting GPU resource availability, but it doesn't shut down existing workloads on the node. When that happens, even the GPU operator tasks themselves cannot be relaunched, as every allocation fails in YK.

Reply:
I just ran ths PR locally against the same cluster that was seeing the issue before, and it's no longer reproducible. I can reproduce quite easily in 1.5.2.

Reply:
> That means allocatedResource could be larger than maxResource, and in that case all allocations can't be fit into that queue even though they don't use the resource type.

The change allows a new allocation to be added ONLY if all the resource types requested in `allocatedResource` fit in the maximum set. It is not a generic option to exceed the maximum. If the queue is already over its maximum for a resource type that is NOT requested the allocation will proceed. Currently that will be blocked.

> If above description is correct, is the situation (allocatedResource is larger than maxResource) acceptable in YK? or it is a temporary state?

No a single allocation can never be larger than the max set on a queue. Even the case that a requested resource type would go over is still blocked.
Examples 1:
- queue max is 10000 vcore, 5 GPU
- queue usage is 8000 vcore, 6 GPU (I lowered the configured max for GPU)
- request is for 1000 vcore

Currently: allocation is blocked (queue is always considered over quota)
New behaviour: allocation is allowed

Examples 2:
- queue max is 10000 vcore, 5 GPU
- queue usage is 8000 vcore, 5 GPU
- request is for 3000 vcore

Current and new behaviour: allocation is blocked queue would be over quota for vcore

Examples 3:
- root queue max is 10000 vcore, 0 GPU (no nodes with GPUs are registered yet
- root queue usage is 8000 vcore, 1 GPU (old GPU job from before the kubelet restart
- request is for 1000 vcore

Currently: all allocations in the system are blocked as the root queue is over quota
New behaviour: allocation is allowed